### DISCUSSION TITLE
Different results in BEPUphysicsint

### DISCUSSION THREAD

**Discussion by dannisliang** [bepu/bepuphysics1#44] [Q&A]
In single-threaded mode, under the conditions of the same number of frames, the same initial coordinates, the same final coordinates, and the same linear velocity, the results of multiple tests show that the final position of the object movement is the same.

However, in the case of different frame numbers and other conditions being the same, the result of multiple tests is that the final position of the object movement is different.

Is this situation related to the different times or different frame numbers.of space.updates? Would you please offer me a solution for this issue? 

Can I record all the data in the physical space of the current frame for reconstruction after the computer restarts? What data needs to be recorded?

Appreciate your help!

**Comment by RossNordby** [comment]
All interactions with the library should be performed in exactly the same order, including with respect to calls to `Space.Update`. Adds, removes, updates, and everything else can all modify internal state which will affect determinism.

To reconstruct a simulation in a deterministic way, you'll need to reproduce all *interactions with the library* in exactly the same way, in exactly the same order.

(Technically, you could try to serialize internal state, but this would be a huge pain and I wouldn't treat it as feasible.)

**Reply by dannisliang** [reply]
What data needs to be recorded?

**Reply by Gavin-Williams** [reply]
I would ask the question - do I really need deterministic physics?

It can get complicated, and it can be impossible. But it depends on what you're trying to do. The answer to your question can really only be determined by you. No one can give you a shopping list of properties to be recorded that would result in deterministic physics. How can they? It's your system. You've constructed the physics environment, the engine flow, the threading model. 

If you really want deterministic physics you will have to work out what data needs to be recorded yourself to make YOUR system work. It's your system, you made. No one else can tell you how your system works.

First, try recording position, linear and angular velocities. And don't forget the state of any objects that may be instantaneously producing forces.

You mention different frame numbers, but other conditions being the same. Are you doing something like variable frame rate, but assuming a fixed framerate for the physics? There must be something different between runs. You'll have to work that out. There may be a bit of trial and error involved.

**Reply by dannisliang** [reply]
Thanks for your advice
